src/Vehicle/Vehicle.cc

```diff
@@ -840,6 +792,70 @@ void Vehicle::_mavlinkMessageReceived(LinkInterface* link, mavlink_message_t mes
         break;
 #endif
     }
+}
+
+void Vehicle::_mavlinkMessageReceived(LinkInterface* link, mavlink_message_t message)
+{
+    // If the link is already running at Mavlink V2 set our max proto version to it.
+    unsigned mavlinkVersion = _mavlink->getCurrentVersion();
+    if (_maxProtoVersion != mavlinkVersion && mavlinkVersion >= 200) {
+        _maxProtoVersion = mavlinkVersion;
+        qCDebug(VehicleLog) << "Vehicle::_mavlinkMessageReceived Link already running Mavlink v2. Setting _maxProtoVersion" << _maxProtoVersion;
+    }
+
+    if (message.sysid != _id && message.sysid != 0) {
+        // We allow RADIO_STATUS messages which come from a link the vehicle is using to pass through and be handled
+        if (!(message.msgid == MAVLINK_MSG_ID_RADIO_STATUS && _containsLink(link))) {
+            return;
+        }
+    }
+
+    if (!_containsLink(link)) {
+        _addLink(link);
+    }
+
+    //-- Check link status
+    _messagesReceived++;
+    emit messagesReceivedChanged();
+    if(!_heardFrom) {
+        if (message.msgid == MAVLINK_MSG_ID_HEARTBEAT) {
+            // This first hearbeat which comes through the vehicle is always the main autopilot component.
+            _heardFrom = true;
+            _mainCompID = message.compid;
+            _messageSeq = message.seq + 1;
+        }
```

Comment:
why relying on this assumption, and not doing the proper mav_autopilot != invalid check?

why not grabbing the opportunity and building up a compid/type table here for use by other code pieces?

Reply:
> why relying on this assumption, and not doing the proper mav_autopilot != invalid check?

Because _mainCompID must be set on the first message since it is used elsewhere. If something changes and the assumption breaks all hell will break lose in other parts of the code as well. So it's way safer to check the assumption. Also this last commit was bad.

Reply:
> why not grabbing the opportunity and building up a compid/type table here for use by other code pieces?

What other code pieces? Nothing else needs it at the moment as far as I can tell.

Reply:
I don't understand this. And I believe it's incorrect.

There is simply no guarantee that the first heartbeat must be the one from the autopilot. Hence in order to determine the autopilots heartbeat you must properly check by

if (message.msgid == MAVLINK_MSG_ID_HEARTBEAT && message.mav_autopilot !=  MAV_AUTOPILOT_INVALID ) 

that's the only safe since conform test. You are of course allowed to add whatever additional test you like on top of this, but adding these additional tests but not the relevant type test doesn't make it way safer.

Also, you actually don't test that the first heartbeat is from the autopilot! In fact, now you not only assume that the first heartbeat is from the autopilot, but even that the very first message from the autopilot is a heartbeat.

If you are worried of bad things happen before _compId is set, you can add an exist before you go to the separate handlers. This would then be safe and would work.

IMHO

:)

Reply:
if one is worried that some flight stacks don't set their heartbeat correct, one could improve the check to 

if (message.msgid == MAVLINK_MSG_ID_HEARTBEAT && (message.mav_autopilot != MAV_AUTOPILOT_INVALID || message.compid == MAV_COMP_ID_AUTOPILOT1) )

:)

Reply:
> There is simply no guarantee that the first heartbeat must be the one from the autopilot.

From a mavlink spec standpoint you are correct. From a QGC standpoint of how MultiVehicleManager (which creates Vehicles) and Vehicle code interact you are incorrect.

Reply:
I agree and confirm that I don't know much about the interaction between MultiVehicleManager and Vehicle code. :)
I wonder however what sort of interaction this could be so that it so vulnerable to what is happening in the vehicle's mavlink receive handling. My feeling is there is then something to improve in architecture for some future version.